File: services/api/src/aec_api/routers/jurisdiction.py
"""R23-JURISDICTION-PACKS — the pack library, and a project's check against the packs for its place.

The library is global (a pack is meant to be shared); the check is project-scoped, because it needs
that project's model and that project's jurisdiction.
"""
from __future__ import annotations

from fastapi import APIRouter, Body, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from .. import jurisdiction_packs as jp
from .. import model_index
from ..db import get_db
from ..deps import get_project
from ..rbac import current_user, require_role

router = APIRouter()


@router.get("/jurisdiction/packs")
def list_packs(jurisdiction: str | None = Query(None), _: str = Depends(current_user)):
    """Every data-requirement pack available, optionally filtered to one jurisdiction.

    The built-in `example` pack asserts nothing about any real place — it is there to show the shape.
    Real packs are imported by someone who can point at the document they came from, which is why
    `authority`, `edition` and `source` are required to store one."""
    try:
        packs = jp.library()
    except jp.PackError as e:
        raise HTTPException(409, str(e)) from e
    if jurisdiction:
        j = jurisdiction.strip().upper()
        packs = [p for p in packs if p.get("jurisdiction") == j]
    return {"packs": packs, "count": len(packs),
            "note": "a pack states whose requirement it is; the built-in `example` pack states that "
                    "it is an example and is attributed to nobody"}


@router.post("/jurisdiction/packs", status_code=201)
def import_pack(pack: dict = Body(...), _: str = Depends(current_user)):
    """Import a data-requirement pack from an authority.

    Refused without `authority`, `edition` and `source`. That is not bureaucracy: a requirement
    nobody can trace is indistinguishable from one somebody made up, and this pack will be used to
    fail other people's models. Selectors are validated with the same parser that will evaluate
    them, so a requirement cannot store cleanly and then silently never match — which reads as a
    pass."""
    try:
        return jp.import_pack(pack)
    except jp.PackError as e:
        raise HTTPException(422, str(e)) from e


@router.delete("/jurisdiction/packs/{pack_id}")
def delete_pack(pack_id: str, _: str = Depends(current_user)):
    try:
        removed = jp.delete_pack(pack_id)
    except jp.PackError as e:
        raise HTTPException(409, str(e)) from e
    if not removed:
        raise HTTPException(404, "no imported pack with that id (the built-in example cannot be deleted)")
    return {"deleted": pack_id}


@router.get("/projects/{pid}/jurisdiction/requirements")
def project_requirements(pid: str, pack: str | None = Query(None),
                         db: Session = Depends(get_db), _: str = Depends(require_role("viewer"))):
    """The data requirements that apply to this project, resolved from its jurisdiction.

    A project with no jurisdiction set gets an empty list and the reason — never a default pack. A
    requirement set from the wrong authority is not a conservative approximation of the right one;
    it is a different answer that looks exactly like it, and it would fail a model against rules
    nobody imposed. Pass `?pack=` to apply one explicitly instead."""
    project = get_project(db, pid)
    try:
        if pack:
            chosen = [p for p in jp.library() if p.get("id") == pack.strip().lower()]
            if not chosen:
                raise HTTPException(404, f"no pack with id {pack!r}")
            return {"jurisdiction": getattr(project, "jurisdiction", None), "packs": chosen,
                    "adopted": True, "explicit": True,
                    "note": "applied by id, not resolved from the project's jurisdiction"}
        return jp.resolve(getattr(project, "jurisdiction", None))
    except jp.PackError as e:
        raise HTTPException(409, str(e)) from e


@router.get("/projects/{pid}/jurisdiction/check")
def project_check(pid: str, pack: str | None = Query(None),
                  db: Session = Depends(get_db), _: str = Depends(require_role("viewer"))):
    """Check the project's model against the data requirements for its jurisdiction.

    Every result carries the pack's authority and edition. A compliance figure detached from whose
    rules it measured is the kind of number that gets quoted at somebody, and `is_example` rides
    along so a run against the demonstration pack can never be mistaken for a real one."""
    project = get_project(db, pid)
    try:
        if pack:
            packs = [p for p in jp.library() if p.get("id") == pack.strip().lower()]
            if not packs:
                raise HTTPException(404, f"no pack with id {pack!r}")
            resolved = {"jurisdiction": getattr(project, "jurisdiction", None), "explicit": True}
        else:
            r = jp.resolve(getattr(project, "jurisdiction", None))
            packs, resolved = r["packs"], {k: r[k] for k in ("jurisdiction", "adopted", "why")}
        if not packs:
            return {**resolved, "model_scored": False, "packs": [],
                    "note": resolved.get("why") or "no pack applies to this project"}
        try:
            model_index.ensure_loaded(pid)
            idx = model_index.get_index(pid)
        except Exception:  # noqa: BLE001 — no model uploaded is an answer, not a failure
            idx = None
        return {**resolved, **jp.check(idx, packs)}
    except jp.PackError as e:
        raise HTTPException(409, str(e)) from e

File: apps/web/src/api/jurisdiction.ts
/** R23-JURISDICTION-PACKS — data requirements that belong to an authority, and the client that
 *  could not reach any of them.
 *
 *  Added 2026-09-11. All **five** routes of this feature had no client caller: the library, the
 *  import and the delete (admin-gated), and the two that CONSUME a pack —
 *  `/projects/{pid}/jurisdiction/requirements` and `/projects/{pid}/jurisdiction/check`. The server
 *  side has been complete since R23: validation that refuses an uncited pack, selector parsing with
 *  the same parser that evaluates it, and a check that carries its own attribution.
 *
 *  The frozen three were visible to `services/api/test_route_reachability.py`. The two consumers
 *  were not, for two different reasons recorded in that file: `check` shares its last segment with
 *  `/projects/{pid}/standards/check`, which the client genuinely calls, and `requirements` collides
 *  with unrelated text. **A leaf is not a name** — which is why the feature read as three-fifths
 *  dark rather than dark.
 */
import { HttpCore } from "./httpCore";

type Ctor<T> = new (...args: any[]) => T;

/** One requirement inside a pack — `rule_library`-shaped, and evaluated by that same engine. */
export interface PackRequirement {
  id: string;
  name: string;
  /** Which elements the requirement is about. */
  scope: string;
  /** What must hold for them. */
  require: string;
  severity: string;
}

/**
 * A data-requirement pack: a named, versioned, ATTRIBUTED set of requirements keyed to a place.
 *
 * `authority`, `edition` and `source` are required by the server, and the reason is the whole point
 * of the feature: this pack will be used to fail somebody else's model, and they have to be able to
 * go and read the rule. A requirement nobody can trace is a house rule wearing a regulator's name.
 */
export interface JurisdictionPack {
  id: string;
  jurisdiction: string;
  authority: string;
  name: string;
  edition: string;
  source: string;
  /** True only for the built-in `example`, which asserts nothing about any real place. */
  is_example?: boolean;
  requirements: PackRequirement[];
}

/** The library: every pack available, optionally narrowed to one jurisdiction. */
export interface PackLibrary { packs: JurisdictionPack[]; count: number; note?: string }

/** Which packs apply to a project, and — when none do — why. */
export interface ProjectRequirements {
  jurisdiction: string | null;
  packs: JurisdictionPack[];
  adopted: boolean;
  /** Set when nothing applies. Null when packs resolved normally. */
  why?: string | null;
  /** True when a pack was applied by id rather than resolved from the project's jurisdiction. */
  explicit?: boolean;
  note?: string;
}

/** One pack's result, carrying the authority whose rules produced the number. */
export interface PackResult {
  id: string | null;
  name: string | null;
  authority: string | null;
  edition: string | null;
  source: string | null;
  is_example: boolean;
  total_rules: number;
  failing_rules: number;
  total_violations: number;
  rules?: { id: string; name?: string; severity?: string; matched?: number; violations?: number }[];
}

/**
 * The check. `model_scored` false means no model was loaded — every figure below is then absent,
 * not zero, and a client that renders 0/0 as "compliant" says the opposite of what happened.
 */
export interface JurisdictionCheck {
  jurisdiction: string | null;
  adopted?: boolean;
  why?: string | null;
  explicit?: boolean;
  model_scored: boolean;
  packs: PackResult[];
  total_requirements: number;
  failing_requirements?: number;
  total_violations?: number;
  satisfied?: boolean;
  attribution?: { pack: string | null; authority: string | null; edition: string | null;
                  is_example: boolean }[];
  note?: string;
}

/** What a delete removed. */
export interface PackDeleted { deleted: string }

export function withJurisdiction<TBase extends Ctor<HttpCore>>(Base: TBase) {
  return class extends Base {
    /** Every data-requirement pack available, optionally filtered to one jurisdiction. */
    jurisdictionPacks(jurisdiction?: string) {
      const q = jurisdiction ? `?jurisdiction=${encodeURIComponent(jurisdiction)}` : "";
      return this.json<PackLibrary>(`/jurisdiction/packs${q}`);
    }
    /**
     * Import a pack from an authority. Admin-only, and refused with a 422 whose `detail` names the
     * missing citation field or the unparseable selector — surface it verbatim rather than a
     * generic failure, because the refusal is the instruction.
     */
    jurisdictionImportPack(pack: unknown) {
      return this.json<JurisdictionPack>("/jurisdiction/packs",
        { method: "POST", body: JSON.stringify(pack) });
    }
    /** Remove an imported pack. The built-in example cannot be deleted (404). */
    jurisdictionDeletePack(packId: string) {
      return this.json<PackDeleted>(`/jurisdiction/packs/${encodeURIComponent(packId)}`,
        { method: "DELETE" });
    }
    /** The packs that apply to this project — resolved from its jurisdiction, or one by id. */
    projectRequirements(pid: string, pack?: string) {
      const q = pack ? `?pack=${encodeURIComponent(pack)}` : "";
      return this.json<ProjectRequirements>(
        `/projects/${encodeURIComponent(pid)}/jurisdiction/requirements${q}`);
    }
    /** Check the project's model against those requirements. */
    jurisdictionCheck(pid: string, pack?: string) {
      const q = pack ? `?pack=${encodeURIComponent(pack)}` : "";
      return this.json<JurisdictionCheck>(
        `/projects/${encodeURIComponent(pid)}/jurisdiction/check${q}`);
    }
  };
}
